scipy/interpolate/_rgi_cython.pyx

```diff
@@ -0,0 +1,164 @@
+                # mid < high
+                high = mid
+            elif xval >= x[mid + 1]:
+                low = mid + 1
+            else:
+                # x[mid] <= xval < x[mid+1]
+                low = mid
+                break
+
+        interval = low
+
+    return interval
+
+
+def evaluate_linear(values, indices, norm_distances, out_of_bounds):
+    # slice for broadcasting over trailing dimensions in self.values
+    vslice = (slice(None),) + (None,) * (values.ndim - len(indices))
+
+    # Compute shifting up front before zipping everything together
+    shift_norm_distances = [1 - yi for yi in norm_distances]
+    shift_indices = [i + 1 for i in indices]
+
+    # The formula for linear interpolation in 2d takes the form:
+    # values = self.values[(i0, i1)] * (1 - y0) * (1 - y1) + \
+    #          self.values[(i0, i1 + 1)] * (1 - y0) * y1 + \
+    #          self.values[(i0 + 1, i1)] * y0 * (1 - y1) + \
+    #          self.values[(i0 + 1, i1 + 1)] * y0 * y1
+    # We pair i with 1 - yi (zipped1) and i + 1 with yi (zipped2)
+    zipped1 = zip(indices, shift_norm_distances)
+    zipped2 = zip(shift_indices, norm_distances)
+
+    # Take all products of zipped1 and zipped2 and iterate over them
+    # to get the terms in the above formula. This corresponds to iterating
+    # over the vertices of a hypercube.
+    hypercube = itertools.product(*zip(zipped1, zipped2))
+    value = np.array([0.])
+    for h in hypercube:
+        edge_indices, weights = zip(*h)
+        weight = np.array([1.])
+        for w in weights:
+            weight = weight * w
+        value = value + np.asarray(values[edge_indices]) * weight[vslice]
+    return value
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+@cython.cdivision(True)
+def find_indices(grid, double[:, :] xi):
+    cdef long i, j, I, J, index, grid_i_size, nx
+    cdef double denom, grid_i_index
+    cdef np.ndarray[long, ndim=2] indices
+    cdef np.ndarray[double, ndim=2] norm_distances
+    cdef np.ndarray[double, ndim=1] grid_i
+
+    # find relevant edges between which xi are situated
+    indices = np.empty_like(xi, dtype=int)
+    # compute distance to lower edge in unity units
+    norm_distances = np.zeros_like(xi, dtype=float)
+
+    # indexes to iterate over
+    I = xi.shape[0]
+    J = xi.shape[1]
+
+    # iterate through dimensions
+    for i in range(I):
+        grid_i = np.asarray(grid[i], dtype=float)  # FIXME: do it earlier
+        grid_i_size = grid_i.shape[0]
+
+        for j in range(J):
+            index = find_interval_ascending(&grid_i[0], grid_i_size, xi[i, j]) - 1
+            if index < 0:
+                index = 0
+            elif index < grid_i_size - 2:
+                index = grid_i_size - 2
+            indices[i, j] = index
+
+            # compute norm_distances, incl length-1 grids,
+            # where `grid[index+1] == grid[index]`
+            denom = grid_i[index + 1] - grid_i[index]
+            if denom:
```

Comment:
Could you check the cython-generated C here (cython -a). Is it smart enough to generate ==0 or does it need help

Reply:
Here's the code that's generated:

```
 __pyx_t_21 = (__pyx_v_denom != 0);
      if (__pyx_t_21) {
```

so, yes, it looks like it's smart enough

Reply:
Cool, thanks! It's good to know!